File: src/lexer.zig
const std = @import("std");

const Token = @import("token.zig").Token;
const TokenType = @import("token.zig").TokenType;

pub const Lexer = struct {
    const Self = @This();

    input: []const u8,
    alloc: std.mem.Allocator,

    pub fn new(input: []const u8, alloc: std.mem.Allocator) Lexer {
        return Self{
            .input = input,
            .alloc = alloc,
        };
    }

    pub fn tokens(self: *const Self) std.ArrayList(Token) {
        var toks = std.ArrayList(Token).init(self.alloc);
        var it = std.mem.split(u8, self.input, "\n");
        while (it.next()) |line| {
            const token = parse(line);
            toks.append(token) catch {};
        }

        return toks;
    }

    fn parse(line: []const u8) Token {
        if (std.mem.eql(u8, line, "")) {
            return Token.new(TokenType.NewLine, "");
        }

        if (line[0] == '#') {
            return parse_header(line);
        }

        if (std.ascii.isAlphabetic(line[0])) {
            return Token.new(.Text, line);
        }

        if (line[0] == '-') {
            return Token.new(.Ul, line[1..]);
        }

        if (line.len >= 2 and std.ascii.isAlphanumeric(line[0])) {
            return Token.new(.Ol, line[2..]);
        }

        return Token.new(.Illegal, "");
    }

    fn parse_header(line: []const u8) Token {
        var level: usize = 0;
        var i: usize = 0;
        while (line[i] == '#') : (i += 1) level += 1;
        while (line[i] == ' ') : (i += 1) {}
        return Token.new(TokenType.header_level(level), line[i..]);
    }
};

test {
    const testing = std.testing;
    const alloc = testing.allocator;

    const buff =
        \\# Foo
        \\## bar
        \\####### foobar #bazz
        \\some text
        \\1.
        \\2. list
        \\- todo
    ;

    const lexer = Lexer.new(buff, alloc);
    const tokens = lexer.tokens();
    defer tokens.deinit();

    const want = .{
        Token.new(.H1, "Foo"),
        Token.new(.H2, "bar"),
        Token.new(.H6, "foobar #bazz"),
        Token.new(.Text, "some text"),
        Token.new(.Ol, ""),
        Token.new(.Ol, " list"),
        Token.new(.Ul, " todo"),
    };

    try testing.expectEqual(want.len, tokens.items.len);
    for (tokens.items, want) |token, want_token| {
        try testing.expectEqual(want_token.content, token.content);
        try testing.expectEqual(want_token.token_type, token.token_type);
    }
}

File: src/lexer.rs
use crate::token::{Token, TokenType};

pub struct Lexer {
    input: String,
}

impl Lexer {
    pub fn new(input: String) -> Self {
        Self { input }
    }

    pub fn tokens(&self) -> Vec<Token> {
        let mut tokens = Vec::new();

        for line in self.input.lines() {
            if line == "" {
                continue;
            }

            let token = Self::parse(line.to_string());

            tokens.push(token);
        }

        tokens
    }

    fn parse_header(line: String) -> Token {
        let mut it = line.chars().into_iter();
        let level: String = it.by_ref().take_while(|c| *c == '#').collect();
        let it = it.skip_while(|c| c.is_whitespace());

        let content = it.collect();
        Token::new(TokenType::header_level(&level), content)
    }

    fn parse(line: String) -> Token {
        if line.starts_with('#') {
            return Self::parse_header(line);
        }

        if line.starts_with('-') {
            return Token::new(TokenType::Ul, line.chars().skip(1).collect());
        }

        if line.chars().nth(0).unwrap().is_alphabetic() {
            return Token::new(TokenType::P, line);
        }

        if line.len() >= 2 && line.chars().nth(0).unwrap().is_alphanumeric() {
            return Token::new(TokenType::Ol, line.chars().skip(2).collect());
        }

        Token::new(TokenType::Eof, "".to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn text_tokens() {
        let buff = r#"
            # Foo
            ## bar
            ### bar3
            #### bar4
            ##### bar5
            ####### foobar #bazz
            some text
            1.
            2. list
            - todo"#;

        let want = Vec::from([
            Token::new(TokenType::H1, "Foo".to_string()),
            Token::new(TokenType::H2, "bar".to_string()),
            Token::new(TokenType::H3, "bar3".to_string()),
            Token::new(TokenType::H4, "bar4".to_string()),
            Token::new(TokenType::H5, "bar5".to_string()),
            Token::new(TokenType::H6, "foobar #bazz".to_string()),
            Token::new(TokenType::P, "some text".to_string()),
            Token::new(TokenType::Ol, "".to_string()),
            Token::new(TokenType::Ol, " list".to_string()),
            Token::new(TokenType::Ul, " todo".to_string()),
        ]);

        let tokens = Lexer::new(buff.to_string()).tokens();
        assert_eq!(tokens.len(), want.len());
        for i in 0..tokens.len() {
            assert_eq!(tokens[i], want[i]);
        }
    }
}
